holopy/inference/noise_model.py

```diff
@@ -83,36 +91,124 @@ def _lnlike(self, pars, data):
+# For now it would be OK since PerfectLensModel only works with single
+# spheres or superpositions, but I'm going to leave this for later.
+class ExactModel(NoiseModel):
+    def __init__(self, scatterer, noise_sd=None, medium_index=None,
+                 illum_wavelen=None, illum_polarization=None, theory='auto',
+                 constraints=[]):
+        super().__init__(scatterer, noise_sd, medium_index, illum_wavelen,
+                         illum_polarization, theory, constraints)
+
+    def forward(self, pars, detector):
+        """
+        Compute a forward model (the hologram)
+
+        Parameters
+        -----------
+        pars: dict(string, float)
+            Dictionary containing values for each parameter used to compute
+            the hologram. Possible parameters are given by self.parameters.
+        detector: xarray
+            dimensions of the resulting hologram. Metadata taken from
+            detector if not given explicitly when instantiating self.
+        """
+        optics, scatterer = self._optics_scatterer(pars, detector)
+        try:
+            return calc_holo(detector, scatterer, theory=self.theory,
+                             scaling=1.0, **optics)
+        except (MultisphereFailure, InvalidScatterer):
+            return -np.inf
+
+
+class PerfectLensModel(NoiseModel):
+    theory_params = ['lens_angle']
+    def __init__(self, scatterer, noise_sd=None, lens_angle=1.0,
+                 medium_index=None, illum_wavelen=None, theory='auto',
+                 illum_polarization=None, constraints=[]):
+        super().__init__(scatterer, noise_sd, medium_index, illum_wavelen,
+                         illum_polarization, theory, constraints)
+        self._use_parameter(lens_angle, 'lens_angle')
+
+    def forward(self, pars, detector):
+        """
+        Compute a forward model (the hologram)
+
+        Parameters
+        -----------
+        pars: dict(string, float)
+            Dictionary containing values for each parameter used to compute
+            the hologram. Possible parameters are given by self.parameters.
+        detector: xarray
+            dimensions of the resulting hologram. Metadata taken from
+            detector if not given explicitly when instantiating self.
+        """
+        optics_kwargs, scatterer = self._optics_scatterer(pars, detector)
+        # We need the lens parameters for the theory:
+        theory_kwargs = self.get_pars(self.theory_params, pars)
+        # FIXME why is self.get_pars a method? It calls like 3 functions
+        # recursively to just do a dictionary indexing on something which
+        # is not even the object's attr.
+        # FIXME would be nice to have access to the interpolator kwargs
+        theory = MieLens(**theory_kwargs)
+        try:
+            return calc_holo(detector, scatterer, theory=theory,
+                             scaling=1.0, **optics_kwargs)
+        except (MultisphereFailure, InvalidScatterer):
```

Comment:
I'd prefer you do it, since I don't know how this works and I'd like the git blame to be on you :) (so people can ask you questions if there are issues).